### ISSUE TITLE
Allow shebang in any file

### ISSUE THREAD

**Issue by mk-pmb** [eslint-community/eslint-plugin-n#336]
### Rule details

I'd like to configure the shebang rule to do all of its regular checks, but not complain "This file needs no hashbang."

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
#!/usr/bin/env -S nodemjs
import main from 'main.mjs';

main(process.argv.slice(2));
```

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

Also the message "This file needs no hashbang." should be a bit more helpful, rather than just patronizing. When I first got the error, my first thought was, "Of course it needs one, how else would my shell know how to run it?" So maybe let's change it to something like "This file contains a hashbang but is not listed as a `bin` entry in package.json."

**Comment by scagood** [comment]
There is an option in the rule for exactly this, it's called [`additionalExecutables`](https://github.com/eslint-community/eslint-plugin-n/blob/master/docs%2Frules%2Fhashbang.md#additionalexecutables)

**Comment by mk-pmb** [comment]
Thanks! I assumed that empty array in the example is an array of glob patterns, so I expressed my intention "any" as `'**'` and that worked. Now instead I need a way to restrict the enforcement of `executableMap` to the files mentioned in the "bin" entry.

**Comment by aladdin-add** [comment]
there are 2 options:

1. additionalExecutables: Mark files as executable that are not referenced by the package.json#bin property, default is `[]`.
2. ignoreUnpublished : Allow for files that are not published to npm to be ignored by this rule, default is `false`.

**Comment by aladdin-add** [comment]
> I have scripts that are meant to be invoked by my node scripts, but would not usually be desirable to invoke from an interactive shell.

it seems an uncommon case; can you share an example to show how it's invoked in the node scripts?

**Comment by mk-pmb** [comment]
Sure. I even factored it into [a separate module](https://github.com/mk-pmb/ubuntu-toast-notifier-pmb-js/) so you can experiment with it. The script invocation boils down to:

> const { execFile } = require('child_process');
> execFile(moduleDir + '/toast.sh', args, EX.spawnOpt);

**Comment by mk-pmb** [comment]
I salvaged the failed example from above by porting a simplified version of the CGI test from my other project. [The test script](https://github.com/mk-pmb/ubuntu-toast-notifier-pmb-js/blob/master/test/cgi-test.mjs) uses cool node.js features (represented by encodeURIComponent) to construct a request for [the CGI](https://github.com/mk-pmb/ubuntu-toast-notifier-pmb-js/blob/master/test/cgi-bin/querytoast.mjs) which uses cool node.js features (represented by decodeURIComponent) to parse its input.

**Comment by aladdin-add** [comment]
I see, thanks! you can safely disable the rule if your `package.json` does not contain `pkg.bin` field.

**Comment by aladdin-add** [comment]
1. the testing files don't have to be published: you can use `pkg.files` or `.npmignore`. https://docs.npmjs.com/cli/v10/configuring-npm/package-json/?v=true#files

2. If these files will not be published, you can use `ignoreUnpublished: true`, in which case none of these files will be reported.

**Comment by mk-pmb** [comment]
In the original project, the CGI is meant to be published because CGI is the preferred way to use the module.

**Comment by mk-pmb** [comment]
> you can safely disable the rule if your package.json does not contain pkg.bin field.

The intent of my rules package is to avoid per-project rule definitions, so disabling the rule in every affected package would run counter to that. Disableing the entire rule would also disable all of its beneficial aspects. Could be a stopgap, but I hope we can do better.

**Comment by aladdin-add** [comment]
the `files` field is to configure which files will be published, please see: https://docs.npmjs.com/cli/v10/configuring-npm/package-json?v=true#files

**Comment by mk-pmb** [comment]
Yes, that is the field that I deduced you probably meant with `pkg.files`. I have the impression that we are on the same page about all facts now, and have a case where none of the supposed solutions above work.

If it helps for motivation, I could make a 1:1 example that includes the `bin` field entries, has various CGIs outside the `test` directory, and explains why it would be useful to have most of these CGIs published to npm. However, I assume everyone here has enough imagination that the current example will suffice.

In my eyes it's now about who wants to take the burden of adding and maintaining a rule for the requested feature.
